src/mlpack/methods/svdplusplus/svdplusplus_function_impl.hpp

```diff
@@ -0,0 +1,611 @@
+{
+  // For an example with rating corresponding to user 'i' and item 'j', the
+  // gradients for the parameters is as follows:
+  //           uservec = v(j) + sum(y(k))
+  //           grad(u(i)) = 2 * (lambda * u(i) - error * v(j))
+  //           grad(v(j)) = 2 * (lambda * v(j) - error * uservec)
+  //           grad(p(i)) = 2 * (lambda * p(i) - error)
+  //           grad(q(j)) = 2 * (lambda * q(j) - error)
+  //           grad(y(k)) = 2 * (lambda * y(k) - error / sqrt(N(u)) * v(j))
+  // 'error' is the prediction error for that example, which is:
+  //           rating(i, j) - p(i) - q(j) - u(i).t() * (v(j) + sum(y(k)))
+  // The full gradient is calculated by summing the contributions over all the
+  // training examples.
+
+  gradient.zeros(rank + 1, numUsers + 2 * numItems);
+
+  for (size_t i = 0; i < data.n_cols; i++)
+  {
+    // Indices for accessing the the correct parameter columns.
+    const size_t user = data(0, i);
+    const size_t item = data(1, i) + numUsers;
+    const size_t implicitStart = numUsers + numItems;
+
+    // Calculate the squared error in the prediction.
+    const double rating = data(2, i);
+    const double userBias = parameters(rank, user);
+    const double itemBias = parameters(rank, item);
+
+    // Iterate through each item which the user interacted with to calculate
+    // user vector.
+    arma::vec userVec(rank, arma::fill::zeros);
+    arma::sp_mat::const_iterator it = implicitData.begin_col(user);
+    arma::sp_mat::const_iterator it_end = implicitData.end_col(user);
+    size_t implicitCount = 0;
+    for (; it != it_end; it++)
+    {
+      userVec += parameters.col(implicitStart + it.row()).subvec(0, rank - 1);
+      implicitCount += 1;
+    }
+    if (implicitCount != 0)
+      userVec /= std::sqrt(implicitCount);
+    userVec += parameters.col(user).subvec(0, rank - 1);
+
+    double ratingError = rating - userBias - itemBias -
+        arma::dot(userVec, parameters.col(item).subvec(0, rank - 1));
+
+    // Gradient is non-zero only for the parameter columns corresponding to the
+    // example.
+    gradient.col(user).subvec(0, rank - 1) +=
+        2 * (lambda * parameters.col(user).subvec(0, rank - 1) -
+        ratingError * parameters.col(item).subvec(0, rank - 1));
+    gradient.col(item).subvec(0, rank - 1) +=
+        2 * (lambda * parameters.col(item).subvec(0, rank - 1) -
+        ratingError * userVec);
+    gradient(rank, user) +=
+        2 * (lambda * parameters(rank, user) - ratingError);
+    gradient(rank, item) +=
+        2 * (lambda * parameters(rank, item) - ratingError);
+    // Calculate gradients for item implicit vector.
+    it = implicitData.begin_col(user);
+    it_end = implicitData.end_col(user);
+    for (; it != it_end; it++)
+    {
+      gradient.col(implicitStart + it.row()).subvec(0, rank - 1) +=
+          2 * (lambda *
+          parameters.col(implicitStart + it.row()).subvec(0, rank - 1) -
```

Comment:
I think `parameters.col(implicitStart + it.row()).subvec(0, rank - 1)` should be normalized i.e. you should divide it by `implicitCount`.